p2p/src/utils/requester/requester.rs

```diff
@@ -118,9 +118,43 @@ impl<E: Clock + GClock + Rng + Metrics, P: PublicKey> Requester<E, P> {
     /// a request is made. This is typically used when a request to the preferred
     /// participant fails.
     ///
-    /// If the request can be made, the participant and request ID are returned. If not,
-    /// the duration to wait before attempting another request is returned.
+    /// Returns `Ok((participant, id))` if a request can be made.
+    /// Returns `Err(duration)` if no participant is available:
+    /// - `Err(Duration::MAX)`: No eligible participants exist (all blocked, excluded, or none
+    ///   registered)
+    /// - `Err(wait_time)`: Participants exist but are rate-limited, retry after `wait_time`
     pub fn request(&mut self, shuffle: bool) -> Result<(P, ID), Duration> {
+        self.request_inner(shuffle, None::<fn(&P) -> bool>)
+    }
+
+    /// Ask for a participant to handle a request, filtered by a predicate.
+    ///
+    /// If `shuffle` is true, the order of participants is shuffled before
+    /// a request is made. This is typically used when a request to the preferred
+    /// participant fails.
+    ///
+    /// Only participants for which `filter` returns `true` will be considered.
+    /// This is useful when the caller knows which specific peers have the data.
+    ///
+    /// Returns `Ok((participant, id))` if a request can be made.
+    /// Returns `Err(duration)` if no participant is available:
+    /// - `Err(Duration::MAX)`: No eligible participants exist (none pass the filter, all blocked,
+    ///   excluded, or none registered)
+    /// - `Err(wait_time)`: Matching participants exist but are rate-limited, retry after
+    ///   `wait_time`
+    pub fn request_filtered(
```

Comment:
I opted to make these two separate methods rather than having `request` take `Option<impl Fn(&P) -> bool>` since when this is `None` the compiler can't infer the type, which would make the API very annoying to use.